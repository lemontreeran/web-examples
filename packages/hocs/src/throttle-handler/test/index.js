import React from 'react'
import { mount } from 'enzyme'

const Target = () => null

describe('throttleHandler', () => {
  let mockJustThrottle = null
  let throttleHandler = null

  beforeEach(() => {
    mockJustThrottle = jest.fn(cb => cb)

    jest.mock('just-throttle', () => mockJustThrottle)
    jest.resetModules()

    throttleHandler = require('../src/').default
  })

  afterAll(() => {
    jest.unmock('just-throttle')
  })

  it('should pass handler arguments through', () => {
    const EnhancedTarget = throttleHandler('testHandler')(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')
    testHandler()
    expect(mockTestHandler.mock.calls).toMatchSnapshot()
  })

  it('should call actual handler during lifecycle', () => {
    const EnhancedTarget = throttleHandler('testHandler')(Target)
    const mockTestHandlerInitial = jest.fn()
    const mockTestHandlerUpdated = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandlerInitial} />
    )
    wrapper.setProps({ testHandler: mockTestHandlerUpdated })
    const testHandler = wrapper.find(Target).prop('testHandler')
    testHandler()
    expect(mockTestHandlerInitial.mock.calls).toMatchSnapshot()
    expect(mockTestHandlerUpdated.mock.calls).toMatchSnapshot()
  })

  it('should call `e.persist()` if it has been passed', () => {
    const EnhancedTarget = throttleHandler('testHandler')(Target)
    const mockTestHandler = jest.fn()
    const mockPersist = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler({ persist: mockPersist })

    expect(mockTestHandler.mock.calls).toMatchSnapshot()
    expect(mockPersist.mock.calls).toMatchSnapshot()
  })

  it('should pass `interval` option to `just-throttle`', () => {
    const EnhancedTarget = throttleHandler('testHandler', 75)(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockJustThrottle.mock.calls).toMatchSnapshot()
  })

  it('should pass `delay` option to `just-throttle` via props', () => {
    const EnhancedTarget = throttleHandler('testHandler', (props) =>
      props.throttle)(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} throttle={75} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockJustThrottle.mock.calls).toMatchSnapshot()
  })

  it('should pass `leadingCall` option to `just-throttle`', () => {
    const EnhancedTarget = throttleHandler('testHandler', 75, true)(Target)
    const mockTestHandler = jest.fn()
    const wrapper = mount(
      <EnhancedTarget testHandler={mockTestHandler} />
    )
    const testHandler = wrapper.find(Target).prop('testHandler')

    testHandler()

    expect(mockJustThrottle.mock.calls).toMatchSnapshot()
  })

  describe('display name', () => {
    const origNodeEnv = process.env.NODE_ENV

    afterAll(() => {
      process.env.NODE_ENV = origNodeEnv
    })

    it('should wrap display name in non-production env', () => {
      process.env.NODE_ENV = 'test'

      const EnhancedTarget = throttleHandler()(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })

    it('should not wrap display name in production env', () => {
      process.env.NODE_ENV = 'production'

      const EnhancedTarget = throttleHandler()(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
