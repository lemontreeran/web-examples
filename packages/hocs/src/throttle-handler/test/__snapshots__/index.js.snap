// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throttleHandler display name should not wrap display name in production env 1`] = `
<ThrottleHandler>
  <Target
    undefined={[Function]}
  />
</ThrottleHandler>
`;

exports[`throttleHandler display name should wrap display name in non-production env 1`] = `
<throttleHandler(Target)>
  <Target
    undefined={[Function]}
  />
</throttleHandler(Target)>
`;

exports[`throttleHandler should call \`e.persist()\` if it has been passed 1`] = `
Array [
  Array [
    Object {
      "persist": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
  ],
]
`;

exports[`throttleHandler should call \`e.persist()\` if it has been passed 2`] = `
Array [
  Array [],
]
`;

exports[`throttleHandler should call actual handler during lifecycle 1`] = `Array []`;

exports[`throttleHandler should call actual handler during lifecycle 2`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`throttleHandler should pass \`delay\` option to \`just-throttle\` via props 1`] = `
Array [
  Array [
    [Function],
    75,
    undefined,
  ],
]
`;

exports[`throttleHandler should pass \`interval\` option to \`just-throttle\` 1`] = `
Array [
  Array [
    [Function],
    75,
    undefined,
  ],
]
`;

exports[`throttleHandler should pass \`leadingCall\` option to \`just-throttle\` 1`] = `
Array [
  Array [
    [Function],
    75,
    true,
  ],
]
`;

exports[`throttleHandler should pass handler arguments through 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;
